include/kernels/staggered_quark_smearing.cuh

```diff
@@ -0,0 +1,222 @@
+          // Forward gather - compute fwd offset for vector fetch
+          const bool ghost
+            = (coord[d] + 2 >= arg.dim[d]) && isActive<kernel_type>(active, thread_dim, d, coord, arg); // 1=>2
+
+          if (doHalo<kernel_type>(d) && ghost) { //?
+
+            const int ghost_idx
+              = ghostFaceIndexStaggered<1>(coord, arg.dim, d, 2); // check nFace=2, requires improved staggered fields
+            const Link U = arg.U(d, coord.x_cb, parity);
+            const Vector in = arg.in.Ghost(d, 1, ghost_idx, their_spinor_parity); //?
+
+            out = mv_add(U, in, out);
+
+          } else if (doBulk<kernel_type>() && !ghost) { // doBulk
+            const int _2hop_fwd_idx = linkIndexP2(coord, arg.dim, d);
+            const Vector in_2hop = arg.in(_2hop_fwd_idx, their_spinor_parity);
+            const Link U_2link = arg.U(d, coord.x_cb, parity);
+            out = mv_add(U_2link, in_2hop, out);
+          }
+        }
+        {
+          // Backward gather - compute back offset for spinor and gauge fetch
+          const bool ghost = (coord[d] - 2 < 0) && isActive<kernel_type>(active, thread_dim, d, coord, arg); // 1=>2
+
+          if (doHalo<kernel_type>(d) && ghost) {
+
+            // when updating replace arg.nFace with 1 here
+            const int ghost_idx
+              = ghostFaceIndexStaggered<0>(coord, arg.dim, d, 2); // check nFace=2, requires improved staggered field
+            const Link U = arg.U.Ghost(d, ghost_idx, parity);
+            const Vector in = arg.in.Ghost(d, 0, ghost_idx, their_spinor_parity);
+
+            out = mv_add(conj(U), in, out);
+
+          } else if (doBulk<kernel_type>() && !ghost) { //?
+
+            const int _2hop_back_idx = linkIndexM2(coord, arg.dim, d);
+            const int _2hop_gauge_idx = _2hop_back_idx;
+
+            const Link U_2link = arg.U(d, _2hop_gauge_idx, parity);
+            const Vector in_2hop = arg.in(_2hop_back_idx, their_spinor_parity);
+            out = mv_add(conj(U_2link), in_2hop, out);
+          }
+        }
+      }
+    }
+  }
+
+  template <int nParity, bool dagger, bool xpay, KernelType kernel_type, typename Arg>
+  struct staggered_qsmear : dslash_default {
+
+    const Arg &arg;
+    constexpr staggered_qsmear(const Arg &arg) : arg(arg) { }
+    static constexpr const char *filename() { return KERNEL_FILE; } // this file name - used for run-time compilation
+
+    template <KernelType mykernel_type = kernel_type>
+    __device__ __host__ inline void operator()(int idx, int s, int parity) // Kernel3D_impl
+    {
+      using real = typename mapper<typename Arg::Float>::type;
+      using Vector = ColorSpinor<real, Arg::nColor, 1>;
+
+      // is thread active (non-trival for fused kernel only)
+      bool active = mykernel_type == EXTERIOR_KERNEL_ALL ? false : true;
+
+      // which dimension is thread working on (fused kernel only)
+      int thread_dim;
+
+      if (arg.is_t0_kernel) {
+        if (arg.t0 < 0) return;
```

Comment:
I'm worried that this `return;` may have unintended consequences, depending on the backend. I'm also worried about undefined behavior if `arg.out(...);` is never assigned, which may happen if this is a case where line 218:
```
if (kernel_type != EXTERIOR_KERNEL_ALL || active) arg.out(coord.x_cb, my_spinor_parity) = out;
```
gets hit.

What do you think @maddyscientist about how this may work with different backends? And @alexstrel @hwancheolJeong can you refactor the logic here in a way to avoid UB in `arg.out`?

Reply:
Regarding your first concern, I am afraid I don't know much about other backends. What can happen on them due to `return;`?

Regarding your second concern, if I understand correctly and my memory is correct, I recall other kernels are defined in a similar way. (e.g., https://github.com/lattice/quda/blob/develop/include/kernels/dslash_staggered.cuh#L206)
And arg.out is assigned here: https://github.com/lattice/quda/blob/feature/quark-smearing.pr/include/kernels/staggered_quark_smearing.cuh#L50.

Reply:
For the latter one, are you concerned when ColorSpinorField instance `out` is created with NULL_FIELD_CREATE option? It can happen in general.. though we are creating "out" with ZERO_FIELD_CREATE in the performTwoLinkGaussianSmearNStep()..

Reply:
And I believe I checked all available dslash policies... (nothing in memory though..)

Reply:
@hwancheolJeong I think @weinbe2's concern with other targets is misplaced, and there's no issue there.  And I agree with you on the second point as well, there's no issue with `arg.out` not being hit here, and this is consistent with what happens in the other dslash kernels.  Marking as resolved.